Add upgrade ('latest') support to pkgng module

jyundt opened the pull request:
##### SUMMARY
Adding support to upgrade FreeBSD packages via the `pkgng` module.

Currently, the module only supports package adds (`present`) and removes (`absent`).



##### ISSUE TYPE

 - Feature Pull Request


##### COMPONENT NAME
pkgng

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (add-pkgng-latest-support 9c76797b8b) last updated 2018/06/16 19:02:14 (GMT -400)
  config file = None
  configured module search path = [u'/home/jyundt/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jyundt/git/ansible/lib/ansible
  executable location = /home/jyundt/git/ansible/bin/ansible
  python version = 2.7.13 (default, Feb  8 2017, 06:57:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]

```


##### ADDITIONAL INFORMATION
I'm open to any suggestions on this PR. I poked around some of the other package modules and wasn't sure if there were any best practices for handling package upgrades.

jyundt committed:
Add upgrade ('latest') support to pkgng module

jyundt committed:
fixup for bad yaml

bcoca commented on lib/ansible/modules/packaging/os/pkgng.py:
```diff
@@ -33,7 +33,7 @@
     state:
         description:
             - State of the package.
```
make note that latest is available since 2.7

jyundt replied:
I realize I may have misunderstood your original request. Did you want me to add `version_added: 2.7` to the `state` option, or just a note in the `state` comments section indicating that `latest` is an option as of 2.7?

bcoca replied:
comment, state option is preexisting, but the new choice is not

jyundt replied:
Ah got it, sorry about that.

Just removed the `version_added:` with c5bb88e.

Lemme know if you'd like any other changes.

bcoca commented on lib/ansible/modules/packaging/os/pkgng.py:
```diff
@@ -105,6 +105,11 @@
     state: absent
+
+# Upgrade package baz
+- pkgng:
```
use `name:` field instead of comment, also add note about 2.7

bcoca commented on lib/ansible/modules/packaging/os/pkgng.py:
```diff
@@ -189,17 +194,35 @@ def install_packages(module, pkgng_path, packages, cached, pkgsite, dir_arg):
+                # rc = 0, no updates available
+                # rc = 1, updates available
+                if old_pkgng:
+                    rc, out, err = module.run_command("%s %s %s upgrade -n -g -U -y %s" % (batch_var, pkgsite, pkgng_path, package))
```
not a show stopper, but the code might be clearer if we just add the '-n' and '%s'/dir_arg in the `if/else` and just execute `run_command` at the end

bcoca commented on lib/ansible/modules/packaging/os/pkgng.py:
```diff
@@ -189,17 +194,35 @@ def install_packages(module, pkgng_path, packages, cached, pkgsite, dir_arg):
 
         if not module.check_mode and not query_package(module, pkgng_path, package, dir_arg):
-            module.fail_json(msg="failed to install %s: %s" % (package, out), stderr=err)
+            module.fail_json(msg="failed to install or upgrade %s: %s" % (package, out), stderr=err)
```
you can use `state` to avoid the 'or' to the user

bcoca commented on lib/ansible/modules/packaging/os/pkgng.py:
```diff
@@ -347,8 +370,8 @@ def main():
 
-    if p["state"] == "present":
-        _changed, _msg = install_packages(module, pkgng_path, pkgs, p["cached"], p["pkgsite"], dir_arg)
+    if p["state"] == "present" or p["state"] == "latest":
```
also not a showstopper, but making for easier reading:
``` if p["state"] in ("present", "latest"):```

jyundt committed:
fixup for bcoca review:
- add `version_added:` to description
- change examples to use `name:` and remove comments
- DRY out install_packages()
- clean up state conditional check in modules.params

jyundt committed:
fixup to remove extra whitespace

jyundt committed:
fixup to handle custom pkgsite when checking for updates

bcoca commented on lib/ansible/modules/packaging/os/pkgng.py:
```diff
@@ -33,9 +33,11 @@
+        choices: [ 'present', 'latest', 'absent' ]
         required: false
         default: present
+        version_added: "2.7"
```
no, state was not added in 2.7

jyundt committed:
fixup to remove misleading `version_added:`
